src/braket/circuits/moments.py

```diff
@@ -138,16 +147,35 @@ def add(self, instructions: Iterable[Instruction]) -> None:
             self._add(instruction)
 
     def _add(self, instruction: Instruction) -> None:
-        qubit_range = instruction.target
-        time = max([self._max_time_for_qubit(qubit) for qubit in qubit_range]) + 1
-
-        # Mark all qubits in qubit_range with max_time
-        for qubit in qubit_range:
-            self._max_times[qubit] = max(time, self._max_time_for_qubit(qubit))
-
-        self._moments[MomentsKey(time, instruction.target)] = instruction
-        self._qubits.update(instruction.target)
-        self._depth = max(self._depth, time + 1)
+        if isinstance(instruction.operator, Noise):
+            qubit_range = instruction.target
+            time = max(0, *[self._max_time_for_qubit(qubit) for qubit in qubit_range])
+
+            # Find the maximum noise_index at the time. The new noise is added to
+            # noise_index = maximum noise_index + 1.
+            max_noise_index = max(
+                [0,]
+                + [
+                    k.noise_index
+                    for k in self._moments
+                    if (k.time == time and isinstance(k, NoiseMomentsKey))
+                ]
+            )
+
+            noise_key = NoiseMomentsKey(time, instruction.target, max_noise_index + 1)
+            self._moments[noise_key] = instruction
```

Comment:
The definition of NoiseMomentKey is to achieve the user experience that adding a noise do not create a new time stamp. In other word, adding noise to a circuit does not change the MomentKey and Instructions of the existing gates. 
Therefore, I use this NoiseMomentKey class which has additional parameter `noise_index` so that a gate and many noises can appear in the same time, as long as their `noise_index` are different.

Reply:
Is it possible to change the type hint of the Moments class so that its key can be a set of types?